analysis/blacklist/blacklist.go

```diff
@@ -0,0 +1,223 @@
+			DisableIdIndex: true,
+		})
+	}
+
+	//create the data
+	//TODO: refactor these into modules
+	bufferSize := 1000
+	buildBlacklistedIPs(uniqueSourceIter, res, ritaBL, bufferSize, true)
+
+	buildBlacklistedIPs(uniqueDestIter, res, ritaBL, bufferSize, false)
+
+	buildBlacklistedHostnames(hostnamesIter, res, ritaBL, bufferSize)
+
+	buildBlacklistedURLs(urlIter, res, ritaBL, bufferSize, "http://")
+
+	//index the data
+	for _, collection := range collections {
+		ensureBLIndexes(ssn, currentDB, collection)
+	}
+
+	ssn.DB(currentDB).C(sourceIPs).EnsureIndex(mgo.Index{
+		Key: []string{"$hashed:ip"},
+	})
+
+	ssn.DB(currentDB).C(destIPs).EnsureIndex(mgo.Index{
+		Key: []string{"$hashed:ip"},
+	})
+	ssn.DB(currentDB).C(hostnames).EnsureIndex(mgo.Index{
+		Key: []string{"$hashed:hostname"},
+	})
+	ssn.DB(currentDB).C(urls).EnsureIndex(mgo.Index{
+		Key: []string{"host", "resource"},
+	})
+}
+
+//ensureBLIndexes ensures the sortable fields are indexed
+//on the blacklist results
+func ensureBLIndexes(ssn *mgo.Session, currentDB, collName string) {
+	ssn.DB(currentDB).C(collName).EnsureIndex(mgo.Index{
+		Key: []string{"conn"},
+	})
+	ssn.DB(currentDB).C(collName).EnsureIndex(mgo.Index{
+		Key: []string{"uconn"},
+	})
+	ssn.DB(currentDB).C(collName).EnsureIndex(mgo.Index{
+		Key: []string{"total_bytes"},
+	})
+}
+
+//buildBlacklists gathers the blacklists to check against
+func buildBlacklists(system *config.SystemConfig) []list.List {
+	//build up the lists
+	var blacklists []list.List
+	//use prebuilt lists
+	if system.BlacklistedConfig.UseIPms {
+		blacklists = append(blacklists, lists.NewMyIPmsList())
+	}
+	if system.BlacklistedConfig.UseDNSBH {
+		blacklists = append(blacklists, lists.NewDNSBHList())
+	}
+	if system.BlacklistedConfig.UseMDL {
+		blacklists = append(blacklists, lists.NewMdlList())
+	}
+	//use custom lists
+	ipLists := buildCustomBlacklists(
+		list.BlacklistedIPType,
+		system.BlacklistedConfig.IPBlacklists,
+	)
+
+	hostLists := buildCustomBlacklists(
+		list.BlacklistedHostnameType,
+		system.BlacklistedConfig.HostnameBlacklists,
+	)
+
+	urlLists := buildCustomBlacklists(
+		list.BlacklistedURLType,
+		system.BlacklistedConfig.URLBlacklists,
+	)
+	blacklists = append(blacklists, ipLists...)
+	blacklists = append(blacklists, hostLists...)
+	blacklists = append(blacklists, urlLists...)
+	return blacklists
+}
+
+//buildCustomBlacklists gathers a custom blacklist from a url or file path
+func buildCustomBlacklists(entryType list.BlacklistedEntryType, paths []string) []list.List {
+	var blacklists []list.List
+	for _, path := range paths {
+		newList := lists.NewLineSeperatedList(
+			entryType,
+			path,
+			86400,
+			tryOpenFileThenURL(path),
+		)
+		blacklists = append(blacklists, newList)
+	}
+	return blacklists
+}
+
+//provide a closure over path to read the file into a line separated blacklist
+func tryOpenFileThenURL(path string) func() (io.ReadCloser, error) {
+	return func() (io.ReadCloser, error) {
```

Comment:
Rita-bl expects a function which returns (io.ReadCloser, error) to get the data from.